### PR TITLE
Loki: Add option to define chunk duration per query

### PR DESCRIPTION
**Pull request by svennergr**
**What is this feature?**

This PR adds the possibility to define chunk ranges per query. 

**Special notes for your reviewer**:

It was not as straight forward as I thought, because you can now have multiple queries with different chunk ranges. I used `groupBy` to now group even further - not just by metrics and logs - but also by their configured chunk ranges.

**How to test**:
Add multiple queries with different chunk ranges.

https://user-images.githubusercontent.com/8092184/225385504-f39198c2-cd9b-4949-bbac-a62f2c1d589e.mov

### CODE REVIEW

**Comment by svennergr** [comment]
@matyax @gabor Let me know if we want to go this direction. If so, I am happy to add some tests to cover the new range groups.

File: public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx [hunk 1]
```diff
@@ -172,6 +172,7 @@ export const LokiQueryEditor = React.memo<LokiQueryEditorProps>((props) => {
           app={app}
           maxLines={datasource.maxLines}
           datasource={datasource}
+          experimentalChunkRange={query.experimentalChunkRange}
```
Comment (matyax) [inline public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx:175]:
```suggestion
          chunkRange={query.chunkRange}
```

File: public/app/plugins/datasource/loki/querybuilder/components/LokiQueryBuilderOptions.tsx [hunk 2]
```diff
@@ -119,6 +132,21 @@ export const LokiQueryBuilderOptions = React.memo<Props>(
               aria-label="Select resolution"
             />
           </EditorField>
+          {config.featureToggles.lokiQuerySplitting && (
+            <EditorField
+              label="Chunk Range"
+              tooltip="Experimental: Defines the range of a single query chunk when query chunking is used."
```
Comment (matyax) [inline public/app/plugins/datasource/loki/querybuilder/components/LokiQueryBuilderOptions.tsx:138]: I don't really think it's necessary to put "experimental" all over the place. If the user is enabling the feature flag, they should be aware of it. Moreover, this makes it look like the range is experimental, while the whole feature is, or hypothetically is.

```suggestion
              tooltip="Defines the range of a single query chunk when query chunking is used."
```

Comment (matyax) [inline public/app/plugins/datasource/loki/querybuilder/components/LokiQueryBuilderOptions.tsx:138]:
```suggestion
              tooltip="Defines the range of a single query chunk when query splitting is used."
```

I need a native speaker to back me up, but "chunking" sounds awful to me.

File: public/app/plugins/datasource/loki/querySplitting.ts [hunk 3]
```diff
@@ -179,19 +173,45 @@ export function runPartitionedQueries(datasource: LokiDatasource, request: DataQ
   const [instantQueries, normalQueries] = partition(queries, (query) => query.queryType === LokiQueryType.Instant);
   const [logQueries, metricQueries] = partition(normalQueries, (query) => isLogsQuery(query.expr));
 
+  const oneDayMs = 24 * 60 * 60 * 1000;
+  const rangePartitionedLogQueries = groupBy(logQueries, (query) =>
+    query.experimentalChunkRange && isValidDuration(query.experimentalChunkRange)
```
Comment (matyax) [inline public/app/plugins/datasource/loki/querySplitting.ts:178]: If `experimentalChunkRange` is not a valid duration, I would stop and return immediately, instead of defaulting to 1 day.

File: public/app/plugins/datasource/loki/querySplitting.ts [hunk 4]
```diff
@@ -179,19 +173,45 @@ export function runPartitionedQueries(datasource: LokiDatasource, request: DataQ
...
+  );
+  const rangePartitionedMetricQueries = groupBy(metricQueries, (query) =>
+    query.experimentalChunkRange && isValidDuration(query.experimentalChunkRange)
+      ? durationToMilliseconds(parseDuration(query.experimentalChunkRange!))
+      : oneDayMs
+  );
+
   const requests: LokiGroupedRequest = [];
-  if (logQueries.length) {
+  for (const [chunkRangeMs, queries] of Object.entries(rangePartitionedLogQueries)) {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/querySplitting.ts:189]: Not really in love of this part.

```suggestion
  for (const chunkRangeMs in rangePartitionedLogQueries) {
```

Being queries: `rangePartitionedLogQueries[chunkRangeMs]`.

**Review by matyax — COMMENTED** [review]
This looks great to me! I would completely remove the `experimental` mention both from the code and descriptions, maybe only keeping the `@experimental` in the js doc.

File: public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx (hunk 1 above)
Comment (svennergr) [reply public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx:175]: I am a bit hesitant to change the variable name, as those are persisted to the dashboard JSON, when you save a panel on a dashboard. So if people now save dashboards with chunking enabled, it might be hard to import those dashboard if we decide to change that property in the future. 

@gabor WDYT?

File: public/app/plugins/datasource/loki/querySplitting.ts (hunk 3 above)
Comment (svennergr) [reply public/app/plugins/datasource/loki/querySplitting.ts:178]: I think it's safe to just remove the check, because we already do it when we set `experimentalChunkRange`: https://github.com/grafana/grafana/blob/svennergr/chunk-range-option/public/app/plugins/datasource/loki/querybuilder/components/LokiQueryBuilderOptions.tsx#L47-L50

Removed in bd3cfbc163.

File: public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx (hunk 1 above)
Comment (gabor) [reply public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx:175]: yeah, that's a legitimate problem.

some options:
1. `chunkRangeExperimental` ? (having `experimental` at the end sounds a little better for me)
2. `testing` or `test` or `trial` instead of `experimental`?
3. keep the current name and add a comment describing the persistence-problem?
4. for the initial run allow changing only in explore-mode, and we can delay this discussion to the future allow-in-dashboard-step 😁 

(in general, would be great to be able to define functions how a `LokiQuery` gets stored and gets loaded back in.. but i don't think we have that currently)

File: public/app/plugins/datasource/loki/types.ts [hunk 5]
```diff
@@ -35,6 +35,12 @@ export interface LokiQuery extends LokiQueryFromSchema {
   // the temporary fix (until this gets improved in the codegen), is to
   // override it here
   queryType?: LokiQueryType;
+
+  /**
+   * This is a property for the experimental query splitting feature.
+   * @experimental
+   */
+  experimentalChunkRange?: string;
```
Comment (gabor) [inline public/app/plugins/datasource/loki/types.ts:43]: could we do the thing where this stores milliseconds-as-number, and when showing it in the `form`, we convert to-and-from a string-representation? WDYT?

(best would be those "nice" string representations like "5m", but for the initial run we could also live with just having people enter the millisecond-value)

File: public/app/plugins/datasource/loki/querySplitting.ts (hunk 4 above)
Comment (svennergr) [reply public/app/plugins/datasource/loki/querySplitting.ts:184]: Could you elaborate why you prefer that?

I don't like `for .. in` because by that you would also iterate over "inherited" properties: https://jsfiddle.net/w1dsunke/

File: public/app/plugins/datasource/loki/types.ts (hunk 5 above)
Comment (svennergr) [reply public/app/plugins/datasource/loki/types.ts:43]: I went with that because I only found `parseDuration` which works `string -> Duration` and `durationToMilliseconds` for `Duration -> number`. I am not sure if we can easily parse a duration string from milliseconds. How would you do that?

I also don't have a big problem with that being a string 😜

Comment (gabor) [reply public/app/plugins/datasource/loki/types.ts:43]: >  I am not sure if we can easily parse a duration string from milliseconds. How would you do that?

efficiently. 😁 


on a more serious note, i'm totally fine with it being a string for the initial version,

but i would strongly object to having it a string in the "production ready" version.. the problem is, the "stability" of that thing (in a dashboard case) depends on the "stability" of the string->number transform-function.. in other words, if that function changes, we suddenly have different values here. in this case we basically outsource the transforming to `parseDuration` and to https://date-fns.org/v2.29.3/docs/add . also, i think this cannot do miliseconds.. so if someone asks "what are the allowed formats of data in this field", we don't really know...

meh, as i said, i'm totally fine with going with string-first, but we should keep in mind that this may have to change later.

Comment (svennergr) [reply public/app/plugins/datasource/loki/types.ts:43]: Well, that would be a problem, but basically not just ours. Things like `timeShift` or `interval` are also saved as strings with exactly the same format:
image

File: public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx (hunk 1 above)
Comment (svennergr) [reply public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx:175]: We could also go with an `experimental` struct inside `LokiQuery`. Something like

```
  experimental: {
    chunkRange: string;
  }
```

Comment (matyax) [reply public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx:175]: The dashboard persistence is a good reason, but again: what would be the difference between having an attribute called `experimentalChunk` or any other name that you think it could change in the future? 
Either if it's called `experimentalSomething` or just `something`, the problem is that it changes, not the experimental part, unless I'm missing anything.

What I don't like about having `experimentalVariable` all over the implementation is that as an internal user it's not giving me any useful information. In `querySplitting.ts` it's the chunk size and has nothing experimental about it. The information seems relevant only for the external user.

File: public/app/plugins/datasource/loki/types.ts (hunk 5 above)
Comment (matyax) [reply public/app/plugins/datasource/loki/types.ts:43]: What I would make sure is that `querySplitting` receives a numeric value, and doesn't have to worry about how it's is stored (string or anything) or presented (duration format). If we need to store it as a duration string it's fine, but I would pass it as a number, otherwise we're leaking external details into the _chunking_ code.

Comment (gabor) [reply public/app/plugins/datasource/loki/types.ts:43]: (nitpick, that is `minInterval` technically 😁 )

Comment (svennergr) [reply public/app/plugins/datasource/loki/types.ts:43]: It's part of the `queries`/`request` object which is being passed to `querySplitting`.

Comment (gabor) [reply public/app/plugins/datasource/loki/types.ts:43]: (another point is, i think we can assume we will never interpret this strings on the server, right?)

but yeah, let's string it for now 😄

Comment (svennergr) [reply public/app/plugins/datasource/loki/types.ts:43]: > (nitpick, that is `minInterval` technically 😁 )

Well, I just looked into a dashboard JSON and that's the first I found 🤷

Comment (svennergr) [reply public/app/plugins/datasource/loki/types.ts:43]: > (another point is, i think we can assume we will never interpret this strings on the server, right?)

Not sure? But even if, Duration strings are common to be interepreted on the server, like we do with our range vectors in metric queries (e.g. the `1h` in `rate({job="\"grafana/data\""} [1h])`)

Comment (matyax) [reply public/app/plugins/datasource/loki/types.ts:43]: > It's part of the queries/request object which is being passed to querySplitting.

Which is an object we have control over and currently transform (fixedRequest). 

https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/datasource.ts#L256-L263

File: public/app/plugins/datasource/loki/querySplitting.ts (hunk 4 above)
Comment (svennergr) [reply public/app/plugins/datasource/loki/querySplitting.ts:184]: Oh interesting, I think `[chunkRangeMs, queries] of Object.entries()` is much more readable.

In this case there are no inherited properties, yes. But that was the reason why I generally don't tend to use `for .. in`.

File: public/app/plugins/datasource/loki/types.ts (hunk 5 above)
Comment (svennergr) [reply public/app/plugins/datasource/loki/types.ts:43]: I see, so should I adjust the type extension [here](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/datasource.ts#L260) and add a number type? But that would also mean I'd need to change the type in `runPartitionedQueries`.

Can do that, but honestly I don't feel we are gaining much.

Comment (matyax) [reply public/app/plugins/datasource/loki/types.ts:43]: Honestly, yes. If we could store it as a number it would fix this for us, but since it's not possible I don't think it's worth going the extra mile for this.

File: public/app/plugins/datasource/loki/querybuilder/components/LokiQueryBuilderOptions.tsx (hunk 2 above)
Comment (matyax) [reply public/app/plugins/datasource/loki/querybuilder/components/LokiQueryBuilderOptions.tsx:138]: It seems that we're rolling with chunking everywhere, so no need to change it.

File: public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx (hunk 1 above)
Comment (svennergr) [reply public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx:175]: Your points are understandable.
Just regarding
> TL;DR the name of the variable doesn't fix the problem of an imported/exported dashboard with an attribute that later changed.
If we now use something like `experimentalChunkRange` with `string` values, but decide to use `number` with our "non-experimental" version, we won't get any problems importing dashboards.

File: public/app/plugins/datasource/loki/querybuilder/components/LokiQueryBuilderOptions.tsx [hunk 6]
```diff
@@ -119,6 +131,21 @@ export const LokiQueryBuilderOptions = React.memo<Props>(
...
+          {config.featureToggles.lokiQuerySplitting && (
+            <EditorField
+              label="Chunk Duration"
+              tooltip="Defines the duration of a single query chunk when query chunking is used."
+            >
+              <AutoSizeInput
+                minWidth={14}
+                type="string"
+                min={0}
+                defaultValue={query.chunkDuration?.toString() ?? '1d'}
```
Comment (gabor) [inline public/app/plugins/datasource/loki/querybuilder/components/LokiQueryBuilderOptions.tsx:143]: do we need the `.toString()` here? it seems to be already a string...

Comment (svennergr) [reply public/app/plugins/datasource/loki/querybuilder/components/LokiQueryBuilderOptions.tsx:143]: Good catch, changed in a4be4d478f.

**Review by matyax — APPROVED** [review]
Great work!
